db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/SelectNode.java

```diff
@@ -1283,6 +1322,9 @@ public ResultSetNode genProjectRestrict(int origFromListSize) throws StandardExc
         if(hasWindows()){
             // Now we add a window result set wrapped in a PRN on top of what we currently have.
             for (WindowNode windowDefinition : windowDefinitionList) {
+                if (childResultColumns.isFromExprIndex()) {
+                    windowDefinition.replaceIndexExpression(childResultColumns);
+                }
```

Comment:
Hmm, you're right, it seems windows and group by columns would refer to child of GBN, not the latest top result columns. Then I wonder why it has been working so far... I added two test cases explicitly test these cases, but they still work.

I then investigated `OrderByList` a bit further. It turns out that in code generation, its expression is not used at all. What matters is `columnPosition`, based on which the generated code knows the indexes of sort keys. This field is assigned in binding phase.

Then I suddenly remember why I bind `OrderByColumns` after replacing index expressions. It's because binding them would resolve the column reference in newly assigned expressions and update their `columnPosition`. `GroupByColumn` doesn't have this issue.

That also explains why replacing index expression late and having wrong column references also work. Because for `OrderByColumn`s, what matters is column references' column position, not their source or chain. We could have a GBN in between and GBN can change the number of result columns, but order by can only refer to the output columns of GBN and that's checked in binding phase. So as long as the order of result columns doesn't change, column position would be correct, despite that the column mapping is not right.

But anyway, I think it's good to make the column mapping right. So I replace `OrderByList` and `WindowDefinition`s together before constructing the `GroupBy` node. Also, I reverted the `replaceIndexExpression()` method for `OrderByList` to still bind it, and added a comment there.